File: src/rover.py
from .grid import Grid
from .directions import Directions

class Rover():

    def __init__(self):
        self.direction = Directions.N
        self.coordinate = {'x': 0, 'y': 0}
        self.command_method_mapping = {
            'R': self.turn_right,
            'L': self.turn_left,
            'M': self.move
        }

    def execute(self, commands):
        for command in commands:
           self.command_method_mapping[command]()
        
        return f"{self.coordinate['x']}:{self.coordinate['y']}:{self.direction.name}"

    def turn_right(self):
        self.turn('R')

    def turn_left(self):
        self.turn('L')

    def turn(self, side):
        next_direction_value = self.direction.value[side]
        self.direction = Directions[next_direction_value]

    def move(self):
        increment = self.direction.value['increment']
        axis = self.direction.value['axis']
        self.coordinate[axis] = self.next_position(self.coordinate[axis], increment)

    def next_position(self, current, increment):
        
        new_position = current + increment

        if new_position > Grid.MAX.value:
            return Grid.MIN.value
        
        if new_position < Grid.MIN.value:
            return Grid.MAX.value
        
        return new_position

File: javascript-typescript/src/rover.ts
const DirectionsMap = {
  N: { L: "W", R: "E", AXIS: "y", INCREMENT: 1 },
  E: { L: "N", R: "S", AXIS: "x", INCREMENT: 1 },
  S: { L: "E", R: "W", AXIS: "y", INCREMENT: -1 },
  W: { L: "S", R: "N", AXIS: "x", INCREMENT: -1 },
};

export class Rover {
  direction: string;
  coordinate: object;
  maxIndex: number;
  minIndex: number;

  constructor() {
    this.direction = "N";
    this.maxIndex = 9;
    this.minIndex = 0;

    this.coordinate = {
      x: this.minIndex,
      y: this.minIndex,
    };
  }

  execute(command: string) {
    for (const char of command) {
      if (char === "M") this.move();
      if (char === "R" || char === "L") this.turn(char);
    }
    return `${this.coordinate['x']}:${this.coordinate['y']}:${this.direction}`;
  }

  private move() {
    const currentDirectionMap = DirectionsMap[this.direction];
    const axis = currentDirectionMap["AXIS"];
    const increment = currentDirectionMap["INCREMENT"];
    const current = this.coordinate[axis];
    const next = this.getNextPosition(current, increment);

    this.coordinate[axis] = next;
  }

  private getNextPosition(current, increment: number) {
    const next = current + increment;
    if (next > this.maxIndex) return this.minIndex;
    if (next < this.minIndex) return this.maxIndex;
    return next;
  }

  private turn(char: string) {
    const currentDirectionMap = DirectionsMap[this.direction];
    this.direction = currentDirectionMap[char];
  }
}
